src/lsp/index.ts

```diff
@@ -0,0 +1,180 @@
+import * as Admzip from "adm-zip";
+import * as which from "which";
+import * as https from "https";
+import * as crypto from "crypto";
+import * as fs from "fs";
+import * as os from "os";
+import * as path from "path";
+import * as vscode from "vscode";
+import { ExtensionContext, WorkspaceConfiguration, workspace } from "vscode";
+import { Executable, LanguageClient, LanguageClientOptions, ServerOptions, TransportKind } from "vscode-languageclient/node";
+
+export abstract class LanguageServerClient {
+  private config: WorkspaceConfiguration;
+  private ls: LanguageClient | null = null;
+  private executableName: string;
+  private context: ExtensionContext;
+  private clientOptions: LanguageClientOptions = {
+    documentSelector: ["meson", { "scheme": "file", language: "meson" }]
+  };
+  abstract repoURL: string;
+  protected abstract get debugExe(): Executable;
+  protected abstract get runExe(): Executable;
+
+  constructor(executableName: string, context: ExtensionContext) {
+    this.executableName = executableName;
+    this.context = context;
+    this.config = workspace.getConfiguration("mesonbuild");
+
+    if (this.languageServerPath == null)
+      return;
+
+    this.startLanguageServer();
+  }
+
+  abstract supportsSystem(os: string, arch: string): boolean
+
+  abstract get downloadInfo(): [url: string, hash: string] | null;
+
+  restart(): void {
```

Comment:
In this function, we want to re-resolve the language server's path.

Reply:
It is doing this automatically, as it  is computed again by referencing `languageServerPath`, if I understand correctly

Reply:
You are correct on this aspect, but I think it would be best to resolve the path in the constructor as mentioned in a previous comment.

Reply:
I don't understand what should be resolved in the constructor. Resolving `which.sync(this.executableName, { nothrow: true })` doesn't make sense, as there is a workflow like:

- Linux user starts VSCode
- Installs LS into /usr/local/bin manually
- Restarts the language server

And because this part is resolved earlier, the language server would still fail to change.

`this.findLanguageServer()` doesn't make sense too:
- Windows/MacOS user starts VSCode
- LS is installed
- `restart()`
- The already resolved result from `findLanguageServer()` is still null

And for `this.config.languageServerPath` I don't see a use case, too.
- User changes config, switches to some other language server
- `restart()`
- The already resolved value would still be the old language server

So what exactly do you mean?